packages/core/src/browser/shell/shell-layout-restorer.ts

```diff
@@ -361,8 +378,17 @@ export namespace ShellLayoutRestorer {
                 pending.push(this.toInflate.pop()!(context));
             }
             await Promise.all(pending);
+
+            if (this.toFilter.length) {
+                this.toFilter.forEach(array => {
+                    const filtered = array.filter(value => value !== undefined);
+                    array.splice(0, array.length, ...filtered);
```

Comment:
Performance: Do this `splice` only if `filtered.length` is smaller than `array.length`

Reply:
> Performance: Do this `splice` only if `filtered.length` is smaller than `array.length`

Interesting, I didn't know about it before. In fact the array will not have so much items to work very slow.
But thinking about performance. To avoid checking for an array length, maybe it's better to reset the array at all and then push all the items..
```
array.length = 0;
array.push(...filtered);
```
WDYT?

Reply:
That might be an alternative to `splice`, but checking `length` for sure has lower cost than resetting the array. And resetting is only necessary if the length has changed.

Another alternative would be this:
```typescript
for (let i = 0; i < array.length; i++) {
    if (array[i] === undefined) {
        array.splice(i--, 1);
    }
}
```
If there are only few undefined elements, this variant might be even better because it does not duplicate the array.

Reply:
In 99% array will not contain undefined elements, so this alternative looks much better than rebuilding the array each time when restoring the layout.
Thanks for this.
I've updated the PR.